### PR TITLE
export: scaling improvement

### PR DESCRIPTION
**Pull request by esq4**
In addition to specifying the exact max.size for export, adds an alternative scaling factor.

Parameters are mutually exclusive. When one changes, the other is cleared and not counted in the export.

In addition, each parameter can be reset by clicking the MMB.

### CODE REVIEW

**Comment by TurboGit** [comment]
I'm not sure about this, let's see if there is supporters for adding this. It seems like cluttering the export for little. What I don't like is that the export size and scaling are exclusive and yet present both on the GUI and clearing each other... Maybe just me.

**Comment by esq4** [comment]
> What I don't like is that the export size and scaling are exclusive and yet present both on the GUI and clearing each other

I'll think about it.

**Comment by esq4** [comment]
If such a variant UI?

image

image

image

The last settings are saving.

**Comment by johnny-bit** [comment]
IMO this is good idea UI-wise now. Seems OK.

For additional points you could throw in dt_calculator_solve in there :) (for both scalling and width/height) or that could be separate PR :)

**Comment by esq4** [comment]
> For additional points you could throw in dt_calculator_solve in there

I was thinking about using a calculator. But this seems redundant to me. And besides that, I tried to make "foolproof".

File: src/common/imageio_module.c
Comment (TurboGit) [inline src/common/imageio_module.c:482]: not space after (

File: src/common/imageio_module.c
Comment (TurboGit) [inline src/common/imageio_module.c:520]: superfluous empty line.

File: src/common/imageio_module.c
Comment (TurboGit) [inline src/common/imageio_module.c:475]: can  you add some comments, not sure I understand all this business:

File: src/libs/export.c
Comment (TurboGit) [inline src/libs/export.c:213]: Can't we have the cale optim routine internal to export.c ? I don't like having that exported as routine of imageio.

File: src/libs/export.c
Comment (TurboGit) [inline src/libs/export.c:313]: widht -> width

Also current dt style is to have all static routines to start with an underscore. I know there is some missing....

File: src/libs/export.c
Comment (TurboGit) [inline src/libs/export.c:342]: doom???

**Review by TurboGit — CHANGES_REQUESTED** [review]
Ok, lot of changes for a first review.

File: src/libs/export.c
Comment (esq4) [reply src/libs/export.c]: I'm sorry, it was a splinter of mental anguish :-(

File: src/common/imageio_module.c
Comment (esq4) [reply src/common/imageio_module.c:485]: This is where the string is cleared for better readability: from "1./2.0" is made "1/2".

File: src/common/imageio_module.h
Comment (TurboGit) [inline src/common/imageio_module.h:234]: either this should have the scale string as parameter (and not use the conf string) and/or a comment should be added to clarify the semantic as as-is it is really not clear what this is about.

File: src/libs/export.c
Comment (esq4) [reply src/libs/export.c]: Yes, it has become possible. Did.

File: src/common/imageio_module.h
Comment (esq4) [reply src/common/imageio_module.h:234]: `The function from the string stored in the "plugins/lighttable/export/resizing_factor" parameter of the configuration file generates the x and y fraction's elements to calculate the scaling factor and improve the readability of the string itself displayed in the "scale" field of the export settings.`

Does this comment clarify anything?

Comment (TurboGit) [reply src/common/imageio_module.h]: Yes, I still don't like this global routine using an hidden prefs. I would prefer having a more generic routine taking a string as parameter and returnin x & y.

**Comment by TurboGit** [comment]
BTW, please rebase on top of current master and do not merge master into your topic branch. TIA.

File: src/libs/export.c
Comment (TurboGit) [inline src/libs/export.c:143]: So my proposal is more something like:

```
gchar *scale_str = dt_conf_get_string(CONFIG_PREFIX "resizing_factor");
dt_imageio_scale_splitup(scale_str, &a, &b);
```

And a, b could probably be renamed num, denum (for numerator denominator).

File: src/common/imageio_module.h
Comment (TurboGit) [reply src/common/imageio_module.h]: You did not change this? Can you explain why? I mean I'd like to understand if there is good arguments to keep it the way it is.

File: src/common/imageio_module.h
Comment (TurboGit) [inline src/common/imageio_module.h:234]: you probably also to fix a & b in comment.

**Comment by TurboGit** [comment]
Apart from my latest comments I tested it and it is working fine on my side. Will be a nice addition.

File: src/common/imageio_module.h
Comment (esq4) [reply src/common/imageio_module.h:234]: Yes, of course I will fix it. I'm sorry.

File: src/libs/export.c
Comment (TurboGit) [reply src/libs/export.c:303]: This is marked as resolved, but it has not been change to const, I suppose this is possible?

File: src/common/imageio_module.h
Comment (TurboGit) [inline src/common/imageio_module.h:238]: Again, I asked to have the scale string as parameter here. I still think this is not ok, using an hidden pref as parameter for a public API sounds wrong.

**Review by TurboGit — CHANGES_REQUESTED** [review]
Last but not least, on #6336 there is a proposal to add a way to select physical units like (inch or centimeter) and a DPI to get a final size in pixels. I suppose this PR will need to be reworked a bit to fit on this new module layout. 

We have a combo names unit with entries : pixel, cm, in

I propose that we add a new entry name : factor

In this mode the label "max size" would be renamed "scaling" with the single entry as done here.

So maybe, just wait a bit for the other PR to be merged as I have asked for some modifications.

File: src/common/imageio_module.h
Comment (esq4) [reply src/common/imageio_module.h:238]: This function is quite specific and works with a single configuration parameter.
The current implementation made it possible to exclude an extra line of accessing the configuration from the code.
Your option returns it. But of course I can implement it.

**Comment by esq4** [comment]
> So maybe, just wait a bit for the other PR to be merged as I have asked for some modifications.

Well I will wait.

**Comment by TurboGit** [comment]
@esq4 : the print-export PR is merged so you can rework this PR on top of master. I would add a new entry in the combobox for support of scaling (with have pixel, cm & inch). And note that this must be only a GUI thing and we want only to have the width & height to be set properly. TIA.

**Comment by esq4** [comment]
OK, wil do.

**Comment by TurboGit** [comment]
Perfect thanks! So I'll wait before merging  #6479 to also document the new scaling support.

**Comment by esq4** [comment]
I offer to do so

image

image

image

image

image

**Comment by TurboGit** [comment]
I don't understand why the scale is not one of the options in the combo where we already have cm, inch, pixel? Having another combo just makes this gui more complex than necessary.

**Comment by esq4** [comment]
> I don't understand why the scale is not one of the options in the combo where we already have cm, inch, pixel? Having another combo just makes this gui more complex than necessary.

"Scale" is a multiplier. It's logical to separate it from units of measurement. In addition, when switching the combo "unit" by scrolling, the sizes are recalculated between in, cm and px. And it seemed convenient to me. Switching to scale (unintentional) in this sequence is confusing.

**Comment by TurboGit** [comment]
I don't see this this was.

The module is an interface to compute the final size in pixels and only the width in pixels and height in pixels are actually needed by this module. Now we have many ways to specify those values:

- enter directly in pixel
- enter in CM and in this case the size in pixels are displayed but not editable
- enter in INCH and in this case the size in pixels are displayed but not editable

I'm proposing to add a new option:

- enter a scale from the size of the picture and in this case the size in pixels should be displayed but not editable.

This way the UI stay not too saturated of controls.

**Comment by esq4** [comment]
How do you suggest calculating the size in pixels when several images of different sizes are selected? I do not know how.

**Comment by TurboGit** [comment]
> And what is the point of displaying the size in pixels when the file is being prepared for printing?

I don't get the question. Why not ? It gives an indication of the actual number of pixels for the exported image as this is the export module (and not the print module) and at the end you have a picture on disk with a number of pixels in width and height.

**Comment by TurboGit** [comment]
> How do you suggest calculating the size in pixels when several images of different sizes are selected? I do not know how.

Indeed, that's a good point. We need to take this into account.

**Comment by TurboGit** [comment]
> Therefore, it is logically correct to separate these methods in UI.

Maybe in the code, but I still don't agree in the GUI. We may want to have an entry scale in the current combo as I proposed and maybe in this case hide the non editable size in pixels. Then in scale mode we want to record the scale but we probably want to do lot of tweaking afterward as IIRC in imageio code which does the export knows only about max pixels sizes.

**Comment by esq4** [comment]
Maybe so?

image

image

image

image

**Comment by TurboGit** [comment]
Yes, looks better to me but it looks like you have removed the non editable size in pixels for cm & inch. This is very informative, I don't see why you've removed this. Also even if not used later for other pictures it may make sense to also keep this for scale display. I have really not strong opinion for this last part...

**Comment by esq4** [comment]
I find it superfluous to display a non-editable size in pixels for several reasons:
- when preparing for printing, only the paper size and printer resolution are important. Maximum size in pixels does not specify anything;
- if the user wants to see the maximum size in pixels, then it is enough to select the appropriate item in the combo;
- with all this, it simply takes up space, saturating of the UI of the module.

There is a thought to display this information in a tooltip, but most likely it will be work for the sake of work and is not in demand by anyone.

**Comment by TurboGit** [comment]
@esq4 : The display of pixel sizes was done by Aurélien some days ago. I won't remove that except if Aurélien approves this change. @aurelienpierre : Can you cast your vote on this one ? Or anyone else BTW...

**Comment by esq4** [comment]
@TurboGit : Yes, it's better to know his opinion.

**Comment by aurelienpierre** [comment]
The actual pixels count is the result of the computation. I just like this kind of double-checking display, both for quick reference and to ensure everything is alright. 

Also, from an user education point of view, it makes very clear that the pixel count is the only thing that actually matters for the export, and DPI × size is merely an interface that produces the pixel count indirectly. Indeed, you will find many users wondering at what DPI they should export for web, or that kind of things that reeks of misunderstanding, so I never let an occasion to use the GUI to teach (because, let's be honest, nobody reads the docs).

**Comment by esq4** [comment]
OK, the approach is clear.

**Comment by esq4** [comment]
image

image

**Comment by TurboGit** [comment]
Some issues:
- The equal pixel displayed in print size (cm or inch) is not refreshed when changing DPI.
- Going from scale to pixels does not refresh the entries (scaling factor is still displayed)
- Select pixel 0 x 0 export OK, then scale 1/2 OK, then select CM and then pixels (just to have the entries selected show 0x0), export and you still have the scale 1/2
- Last but not least, each time I export I have on the console:

```
[lcms2] error 9: Wrong output color space on transform
unsupported output profile `ISO Coated' has been replaced by sRGB!
[lcms2] error 9: Wrong output color space on transform
unsupported output profile `ISO Coated' has been replaced by sRGB!
[lcms2] error 9: Wrong output color space on transform
unsupported output profile `ISO Coated' has been replaced by sRGB!
[lcms2] error 9: Wrong output color space on transform
unsupported output profile `ISO Coated' has been replaced by sRGB!
[export_job] exported to `/home/obry/Images/Photos/export/tests/img_19_10.jpg'
```

Please be sure to test all possible combinations. Thanks.

**Comment by TurboGit** [comment]
Hum, looks like last issue is not due to this PR!

**Comment by esq4** [comment]
> Some issues:
> ...

Thanks, fixed. I have not found others yet.

**Comment by TurboGit** [comment]
Still some issues:
- select pixels 1920x1920 export OK
- select scale (let 1) as proposed => export image is 1920 WRONG
- enter 1 in scale (already there but editing seems to reset)
- export OK - size is max
- back to pixels, the size has been reset to 0 x 0 (would be nice to keep last entered 1920x1920).

**Review by TurboGit — CHANGES_REQUESTED** [review]
Still some issues.

**Comment by esq4** [comment]
> * enter 1 in scale (already there but editing seems to reset)

Yes, it is reset only when editing. When switching, I do not change, since it is very easy to do it through negligence.
I'll think about how to solve this.

**Comment by esq4** [comment]
> > * enter 1 in scale (already there but editing seems to reset)
> 
> Yes, it is reset only when editing. When switching, I do not change, since it is very easy to do it through negligence.
> I'll think about how to solve this.

Removed reseting values when switching. All settins are saving.
Added resetting of dimensions by MMB.

File: src/libs/export.c
Comment (TurboGit) [inline src/libs/export.c:1149]: should be "to cancel..." but I would just drop this line. I would rephrase the line above:

"click middle mouse button to reset to 0"

File: src/libs/export.c
Comment (TurboGit) [inline src/libs/export.c:1153]: see above

File: src/libs/export.c
Comment (TurboGit) [inline src/libs/export.c:1170]: see above

File: src/libs/export.c
Comment (TurboGit) [inline src/libs/export.c:1174]: see above

File: src/libs/export.c
Comment (TurboGit) [inline src/libs/export.c:1203]: "click the middle mouse button to reset to 1."

**Review by TurboGit — CHANGES_REQUESTED** [review]
All working fine on my side, just some comments about the English text in tooltip.

File: src/libs/export.c
Comment (esq4) [reply src/libs/export.c:1149]: I tried to explain the meaning of this operation. Think don't need it?

Comment (TurboGit) [reply src/libs/export.c:1149]: Yes I think it is not needed and not proper English.

**Review by TurboGit — APPROVED** [review]
Looks good to me now. I'll merge now to get more field testing but on my side it is working properly now. Thanks a lot !

**Pull request merged**
